independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/BootstrapAppModelFactory.java

```diff
@@ -0,0 +1,512 @@
+                        }
+                    } catch (IOException e) {
+                        log.warn("Failed to read deployment classpath cache from " + cachedCpPath + " for "
+                                + localProject.getAppArtifact(), e);
+                    }
+                }
+            }
+
+            AppModelResolver appModelResolver = getAppModelResolver();
+            CurationResult curationResult = new CurationResult(appModelResolver
+                    .resolveModel(localProject.getAppArtifact()), new FixedSuppler<>(appModelResolver));
+            if (cachedCpPath != null) {
+                Files.createDirectories(cachedCpPath.getParent());
+                try (DataOutputStream out = new DataOutputStream(Files.newOutputStream(cachedCpPath))) {
+                    out.writeInt(CP_CACHE_FORMAT_ID);
+                    out.writeInt(localProject.getWorkspace().getId());
+                    ObjectOutputStream obj = new ObjectOutputStream(out);
+                    obj.writeObject(curationResult.getAppModel());
+                } catch (Exception e) {
+                    log.warn("Failed to write classpath cache", e);
+                }
+            }
+            return curationResult;
+        } catch (Exception e) {
+            throw new BootstrapException("Failed to create the application model for " + localProject.getAppArtifact(), e);
+        }
+    }
+
+    /**
+     * If no maven project is around do discovery based on the class path.
+     *
+     * This is used to run gradle tests, and allows them to run from both the IDE
+     * and the gradle test task
+     *
+     */
+    private CurationResult doClasspathDiscovery() {
+        try {
+            AppModelResolver resolver = getAppModelResolver();
+
+            Set<URL> urls = new HashSet<>();
+            //this is pretty yuck, but under JDK11 the URLClassLoader trick does not work
+            Enumeration<URL> manifests = Thread.currentThread().getContextClassLoader().getResources("META-INF/MANIFEST.MF");
+            while (manifests.hasMoreElements()) {
+                URL url = manifests.nextElement();
+                if (url.getProtocol().equals("jar")) {
+                    String path = url.getPath();
+                    if (path.startsWith("file:")) {
+                        path = path.substring(5, path.lastIndexOf('!'));
+                        urls.add(new File(URLDecoder.decode(path, StandardCharsets.UTF_8.name())).toURI().toURL());
+                    }
+                }
+            }
+            List<AppDependency> artifacts = new ArrayList<>();
+            for (URL jarUrl : urls) {
+                try (JarInputStream file = new JarInputStream(jarUrl.openConnection().getInputStream())) {
+                    JarEntry entry = file.getNextJarEntry();
+                    while (entry != null) {
+                        if (entry.getName().endsWith("/pom.properties") && entry.getName().startsWith("META-INF/maven")) {
```

Comment:
The issue is that this is not very reliable. Neither `pom.properties` nor `pom.xml` are always present. I don't think JARs built with Gradle contain those. You can try a simple test: scan JARs in your local repo and list those that don't include `META-INF/maven`.
So this approach can be used only as "the best effort" in cases when nothing else is available.

Reply:
Thinking about it we don't actually need the artifact names here, the path is what is really important. We just need some way to make sure we don't get jdk jars.

Reply:
Right, unless we wanted to apply a filter.